certora/specs/Tokens.spec

```diff
@@ -0,0 +1,137 @@
+    function _.borrowRate(MetaMorphoHarness.MarketParams, MetaMorphoHarness.Market) external => NONDET;
+}
+
+function summaryIdToMarketParams(MetaMorphoHarness.Id id) returns MetaMorphoHarness.MarketParams {
+    MetaMorphoHarness.MarketParams marketParams;
+    uint256 lastUpdated = Morpho.lastUpdate(id);
+
+    // Safe require because markets in the supply/withdraw queue have positive last update (see LastUpdated.spec).
+    require lastUpdated > 0;
+    // Safe require because it is a verified invariant in Morpho Blue.
+    require lastUpdated > 0 => Morpho.libId(marketParams) == id;
+
+    return marketParams;
+}
+
+function summarySupply(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) returns (uint256, uint256) {
+    assert shares == 0;
+    assert onBehalf == currentContract;
+    assert data.length == 0;
+
+    // Safe require because it is a verified invariant.
+    require hasSupplyCapIsEnabled(Morpho.libId(marketParams));
+    // Safe require because it is a verified invariant.
+    require isEnabledHasConsistentAsset(marketParams);
+
+    // Summarize supply as just a transfer for the purpose of this specification file, which is sound because only the properties about tokens are verified in this file.
+    Util.safeTransferFrom(marketParams.loanToken, currentContract, MORPHO(), assets);
+
+    return (assets, shares);
+}
+
+function summaryWithdraw(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) returns (uint256, uint256) {
+    assert onBehalf == currentContract;
+    assert receiver == currentContract;
+
+    MetaMorphoHarness.Id id = Morpho.libId(marketParams);
+    uint256 withdrawn = Util.withdrawnAssets(MORPHO(), id, assets, shares);
+
+    bool enabled;
+    _, enabled, _ = config(id);
+    // Safe require because:
+    // - for reallocate this is checked in the code
+    // - for withdraw, it is verified that markets in the withdraw queue are enabled.
+    require enabled;
+    // Safe require because it is a verified invariant.
+    require isEnabledHasConsistentAsset(marketParams);
+
+    // Summarize supply as just a transfer for the purpose of this specification file, which is sound because only the properties about tokens are verified in this file.
+    Util.safeTransferFrom(marketParams.loanToken, MORPHO(), currentContract, withdrawn);
+
+    return (withdrawn, shares);
+}
+
+// Check that, on deposit, MetaMorpho's balance does not change and that Morpho's balance increases by the corresponding amount.
+rule depositTokenChange(env e, uint256 assets, address receiver) {
+    address asset = asset();
+    address morpho = MORPHO();
+
+    // Trick to require that all the following addresses are different.
+    require morpho == 0x10;
+    require asset == 0x11;
+    require currentContract == 0x12;
+    require e.msg.sender == 0x13;
+    require receiver == 0x14;
```

Comment:
it should not be an issue that the asset and receiver are the same address as the sender, mm, or blue

Reply:
Let me try it out, I think you are right and we can remove some of those. I had an issue when the asset was MM, which sounds like a case that would not happen so I did not bother looking too much into it

Reply:
`receiver` removed.

`asset` cannot be equal to MM though, because the fee (on MM) can mint new shares to the fee recipient which would increase his balance. Then we would need to require that the fee recipient is someone else, but it is just more complicated than the current rule